### PR TITLE
FdoSecrets: Add unit tests for secret service integration

### PR DESCRIPTION
**Pull request by Aetf**
[TIP]:  # ( Provide a general summary of your changes in the title above ^^ )

## Type of change
[NOTE]: # ( Please remove all lines which don't apply. )
- ✅ Refactor (significant modification to existing code)
- ✅ New feature (non-breaking change which adds functionality)

## Description and Context
[NOTE]: # ( Describe your changes in detail, why is this change required? )
[NOTE]: # ( Describe the context of your change. Explain large code modifications. )
[NOTE]: # ( If it fixes an open issue, please add "Fixes #XXX" as necessary )
This adds proper unit tests to the secret service integration plugin. The tests cover most exposed API and some regression tests for previously fixed issues.

The first commit adds all test cases. The whole test is added as a GUI test because currently the plugin still depends on `DatabaseWidget`. Other changes included in this commit:

* The ownership management of `DatabaseOpenDialog` inside `DatabaseTabWidget` is changed from using `QScopedPointer` to setting `QObject` parent. The lifetime of `DatabaseOpenDialog` is not affected, but it enables the testing code to access the `DatabaseOpenDialog` via object hierarchy.
* Added a new argument `LAUNCHER` to the `add_unit_test` macro. The calling site can specify a launcher program used to start the test binary. This is needed to use `dbus-run-session` to run the dbus test.

The following commits fix issues discovered by newly added tests. One notable change is the addition of `skipProtected` option to the entry searcher. The option is disabled by default, so the behavior remains the same as before. When enabled, search terms referring to password or protected attributes are filtered out before the searcher performs the match. And if there are no terms left after filtering, the default is changed to **not** match, so the search API via DBus won't return everything when passing only this kind of terms. This is a more general solution of #4008, which only considers passwords.

## Testing strategy
[NOTE]: # ( Please describe in detail how you tested your changes. )
[TIP]:  # ( We expect new code to be covered by unit tests and documented with doc blocks! )
`make test`

CTest runner will automatically use `dbus-run-session` to run `testguifdosecrets`.

## Checklist:
[NOTE]: # ( Please go over all the following points. )
[NOTE]: # ( Again, remove any lines which don't apply. )
[NOTE]: # ( Pull Requests that don't fulfill all [REQUIRED] requisites are likely )
[NOTE]: # ( to be sent back to you for correction or will be rejected.  )
- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**
- ✅ My change requires a change to the documentation, and I have updated it accordingly.
- ✅ I have added tests to cover my changes.

### CODE REVIEW

**Comment by Aetf** [comment]
Enough fighting CI for today :man_facepalming:. The last issue seems to be missing dbus-run-session in the CI environment.

**Comment by droidmonkey** [comment]
Easy fix, we can add that. Are there any version contraints?

**Comment by Aetf** [comment]
Just did a quick search, on ubuntu, `dbus-run-session` is provided by the `dbus` package. I think the current [1.10.6](https://packages.ubuntu.com/xenial/dbus) available in ubuntu repo should be fine, although I've only tested with 1.12.6.

**Comment by Aetf** [comment]
Ping. I rebased this on the latest `develop`.

**Comment by droidmonkey** [comment]
If the tests pass I'll merge it :-D

**Comment by Aetf** [comment]
Cool! Thanks for reviewing. This depends on #4776 to correctly cleanup between test cases, though.

**Comment by droidmonkey** [comment]
Merged that one, please rebase again

**Comment by Aetf** [comment]
The failure in CI seems unrelated in `TestCli::testEdit` (Windows) and `TestCli::testYubiKeyOption` (Ubuntu).

**Comment by droidmonkey** [comment]
Yah sometimes those trip for random reasons

File: src/gui/MainWindow.cpp [hunk 1]
```diff
@@ -176,7 +176,7 @@ MainWindow::MainWindow()
 #endif
 
 #ifdef WITH_XC_FDOSECRETS
-    auto fdoSS = new FdoSecretsPlugin(m_ui->tabWidget);
+    auto fdoSS = new FdoSecretsPlugin(m_ui->tabWidget, this);
```
Comment (droidmonkey) [inline src/gui/MainWindow.cpp:179]: Don't pass MainWindow, use the `getMainWindow()` function.

File: src/gui/DatabaseTabWidget.cpp [hunk 2]
```diff
@@ -49,7 +49,7 @@ DatabaseTabWidget::DatabaseTabWidget(QWidget* parent)
     : QTabWidget(parent)
     , m_dbWidgetStateSync(new DatabaseWidgetStateSync(this))
     , m_dbWidgetPendingLock(nullptr)
-    , m_databaseOpenDialog(new DatabaseOpenDialog())
+    , m_databaseOpenDialog(new DatabaseOpenDialog(this))
```
Comment (droidmonkey) [inline src/gui/DatabaseTabWidget.cpp:52]: Why are you changing this?

File: src/gui/DatabaseTabWidget.h [hunk 3]
```diff
@@ -108,7 +108,7 @@ private slots:
 
     QPointer<DatabaseWidgetStateSync> m_dbWidgetStateSync;
     QPointer<DatabaseWidget> m_dbWidgetPendingLock;
-    QScopedPointer<DatabaseOpenDialog> m_databaseOpenDialog;
+    QPointer<DatabaseOpenDialog> m_databaseOpenDialog;
```
Comment (droidmonkey) [inline src/gui/DatabaseTabWidget.h:111]: Why was this changed?

File: src/core/EntrySearcher.cpp [hunk 4]
```diff
@@ -136,11 +137,21 @@ void EntrySearcher::setCaseSensitive(bool state)
...
+void EntrySearcher::setSkipProtected(bool skip)
+{
+    m_skipProtected = skip;
+}
+
+bool EntrySearcher::isSkipProtected() const
+{
+    return m_skipProtected;
+}
+
```
Comment (droidmonkey) [inline src/core/EntrySearcher.cpp:154]: Either set it in the constructor (persistent) or via function (non-persistent), but don't support both. EntrySearcher is meant to be one-time-use object. Recommend removing these.

File: src/core/EntrySearcher.cpp [hunk 5]
```diff
@@ -150,8 +161,26 @@ bool EntrySearcher::searchEntryImpl(const Entry* entry)
...
+            }
+            terms << term;
+        }
+        // empty search terms match nothing
+        if (terms.isEmpty()) {
+            return false;
+        }
+    } else {
+        terms = m_searchTerms;
+    }
```
Comment (droidmonkey) [inline src/core/EntrySearcher.cpp:180]: This filtering should be done when the search terms are parsed.

File: src/core/EntrySearcher.h [hunk 6]
```diff
@@ -62,13 +62,21 @@ class EntrySearcher
     QList<Entry*> repeatEntries(const QList<Entry*>& entries);
 
     void setCaseSensitive(bool state);
-    bool isCaseSensitive();
+    bool isCaseSensitive() const;
+
+    /**
+     * Set the next search to skip protected attributes or not
+     * @param skip
+     */
```
Comment (droidmonkey) [inline src/core/EntrySearcher.h:70]: No docstrings in header files.

File: src/gui/DatabaseTabWidget.cpp (hunk 2 above)
Comment (Aetf) [reply src/gui/DatabaseTabWidget.cpp:52]: Because otherwise, the testing code has no way to access the database open dialog. And changing from using a `QScopedPointer` to using `QObject` parent should not change the lifetime of `DatabaseOpenDialog` in any non-negligible way.

File: src/gui/DatabaseTabWidget.h (hunk 3 above)
Comment (Aetf) [reply src/gui/DatabaseTabWidget.h:111]: To enable access in testing code. See the other comment.

File: src/gui/DatabaseTabWidget.cpp (hunk 2 above)
Comment (droidmonkey) [reply src/gui/DatabaseTabWidget.cpp:52]: Don't change the ABI for testing purposes. We have that in other places and I removed a bunch of it for 2.6. You can get the current active window (the dialog in this case) with `qApp->activeWindow()`.

**Comment by leifkotsch** [comment]
🐈

File: src/gui/DatabaseTabWidget.cpp (hunk 2 above)
Comment (droidmonkey) [reply src/gui/DatabaseTabWidget.cpp:52]: Sorry I was wrong with this one, you are correct the ABI did not change.

File: src/gui/DatabaseTabWidget.h (hunk 3 above)
Comment (droidmonkey) [reply src/gui/DatabaseTabWidget.h:111]: Good to go here too

File: src/core/EntrySearcher.cpp (hunk 4 above)
Comment (droidmonkey) [reply src/core/EntrySearcher.cpp]: I made the necessary changes

**Comment by droidmonkey** [comment]
I made some minor tweaks (see comments above), good to go now.

File: src/core/EntrySearcher.cpp [hunk 7]
```diff
@@ -161,26 +151,8 @@ bool EntrySearcher::searchEntryImpl(const Entry* entry)
...
-        }
-        // empty search terms match nothing
-        if (terms.isEmpty()) {
-            return false;
-        }
-    } else {
-        terms = m_searchTerms;
-    }
-
     bool found;
```
Comment (Aetf) [inline src/core/EntrySearcher.cpp]: `found` may be uninitialized if every term in `m_searchTerms` are protected and `m_skipProtected == true`. However, neither initializing it to `true` or `false` works in all cases.

Consider a database with two entries:
* entry 1: title "Apple", password "pass"
* entry 2: title "Banana", password "word"

If `found` initializes to `true`, a search query "password:pass" will return entry 1, while it should return nothing

If `found` initializes to `false`, a search query "password:pass title:Apple" returns nothing, but it should return entry 1.

Basically, all terms matching protected attributes should be removed first to not affecting the shortcut logic.

Comment (droidmonkey) [reply src/core/EntrySearcher.cpp]: Good point, then we should continue instead of break to skip the short circuit logic and default found to false.

Comment (Aetf) [reply src/core/EntrySearcher.cpp]: I split the logic of `term.exclude` handling and short circuit. It's now cleaner IMHO especially with the addition of `skipProtected` to the mix.

I also added my above comment as a new test case.

**Comment by Aetf** [comment]
Rebased and squashed the fixups into the original commit.

**Comment by Aetf** [comment]
It's `TestCli::testYubiKeyOption` again failing on Windows :/
